src/domain/ledger/imbalance-calculator.ts

```diff
@@ -0,0 +1,50 @@
+const MS_PER_HOUR = (60 * 60 * 1000) as MilliSeconds
+const MS_PER_DAY = (24 * MS_PER_HOUR) as MilliSeconds
+
+export const ImbalanceCalculator = ({
+  volumeLightningFn,
+  volumeOnChainFn,
+  sinceDaysAgo,
+}: ImbalanceCalculatorConfig): ImbalanceCalculator => {
+  const since = new Date(new Date().getTime() - sinceDaysAgo * MS_PER_DAY)
+
+  const getImbalance = async ({
+    volumeFn,
+    walletId,
+    since,
+  }: {
+    volumeFn: GetVolumeSinceFn
+    walletId: WalletId
+    since: Date
+  }) => {
+    const volume_ = await volumeFn({
+      walletId,
+      timestamp: since,
+    })
+    if (volume_ instanceof Error) return volume_
+
+    return (volume_.incomingBaseAmount - volume_.outgoingBaseAmount) as Imbalance
+  }
+
+  const getSwapOutImbalance = async (walletId: WalletId) => {
+    const lnImbalance = await getImbalance({
+      since,
+      walletId,
+      volumeFn: volumeLightningFn,
+    })
+    if (lnImbalance instanceof Error) return lnImbalance
+
+    const onChainImbalance = await getImbalance({
+      since,
+      walletId,
+      volumeFn: volumeOnChainFn,
+    })
+    if (onChainImbalance instanceof Error) return onChainImbalance
+
+    return (lnImbalance - onChainImbalance) as Imbalance
```

Comment:
> Some users will have outsized effects on the bank’s loop imbalance. For example, a merchant that routinely receives offchain payments, and periodically withdraws bitcoin from their account onchain, will impart a consistently negative rate of change on the bank’s loop imbalance. Unless this usage is balanced by users with opposite effects on the bank’s loop imbalance, the bank will eventually have to swap out to rebalance its onchain and offchain liquidity and regain some inbound liquidity.

It doesn't seem like there is any additional fee for this case as-is though? Bc when you calculate the dynamic fee:

```js
const aboveThreshold = Math.max(imbalance - thresholdImbalance, 0)
```

Any `imbalance` that is negative (as in the case quoted above) `aboveThreshold` will default to `0`. Maybe a `Math.abs(imbalance)` is needed in `WithdrawalFeeCalculator`?

Reply:
the user get higher fees only in the case of positive (swap out) imbalance.

swap out means the net inbound flow for lightning is more than net inbound flow for onchain (ie: user send funds on lightning and withdraw it onchain). 

if a user does the opposite then this is great, and there is no fees for that. 

the formula make it so that until the imbalance reach `thresholdImbalance`, there is no additional fees. `aboveThreshold` would be 0 with the formula above.